metaserver/src/main/scala/scala/meta/languageserver/mtags/ScalaMtags.scala

```diff
@@ -10,26 +10,29 @@ object ScalaMtags {
       override def language: String =
         "Scala212" // TODO(olafur) more accurate dialect
       override def indexRoot(): Unit = apply(root)
-      override def apply(tree: Tree): Unit = {
-        val old = currentOwner
-        val next = tree match {
-          case t: Source => Continue
-          case t: Template => Continue
-          case t: Pkg => pkg(t.ref); Continue
-          case t: Pkg.Object => term(t.name, PACKAGEOBJECT); Continue
-          case t: Defn.Class => tpe(t.name, CLASS); Continue
-          case t: Defn.Trait => tpe(t.name, TRAIT); Continue
-          case t: Defn.Object => term(t.name, OBJECT); Continue
-          case t: Defn.Def => term(t.name, DEF); Stop
-          case Defn.Val(_, Pat.Var(name) :: Nil, _, _) => term(name, DEF); Stop
-          case Defn.Var(_, Pat.Var(name) :: Nil, _, _) => term(name, DEF); Stop
-          case _ => Stop
+      override def apply(tree: Tree): Unit = withOwner() {
+        def continue(): Unit = super.apply(tree)
+        def stop(): Unit = ()
+        tree match {
+          case t: Source => continue()
+          case t: Template => continue()
+          case t: Pkg => pkg(t.ref); continue()
+          case t: Pkg.Object => term(t.name, PACKAGEOBJECT); continue()
+          case t: Defn.Class => tpe(t.name, CLASS); continue()
+          case t: Defn.Trait => tpe(t.name, TRAIT); continue()
+          case t: Defn.Object => term(t.name, OBJECT); continue()
+          case t: Defn.Def => term(t.name, DEF); stop()
+          case t: Decl.Def => term(t.name, DEF); stop()
+          case Defn.Val(_, Pat.Var(name) :: Nil, _, _) =>
+            term(name, VAL); stop()
+          case Decl.Val(_, Pat.Var(name) :: Nil, _) =>
+            term(name, VAL); stop()
+          case Defn.Var(_, Pat.Var(name) :: Nil, _, _) =>
+            term(name, VAR); stop()
+          case Decl.Var(_, Pat.Var(name) :: Nil, _) =>
+            term(name, VAR); stop()
```

Comment:
I think `Decl.Type` and `Defn.Type` are missing. 
Btw, what are other cases of `Decl.Type` besides abstract type members?

Reply:
Another thing I forgot to mention: I think this doesn't handle things like `val (x, y) = ...`, because you restrict to a single `Pat.Var` name. How about changing it to something like

```scala
case Decl.Var(_, patvars, _) =>
  patvars.collect { case Pat.Var(name) => term(name, VAR) }; stop()
```